### PR TITLE
Delivery for Tomte

### PR DESCRIPTION
**Pull request by fevac**
### Added

- Added delivery of files for tomte

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b tomte_delivery -a
    ```

### How to test

Tomte:
- [x] nothing happens when data delivery is `no delivery`:`cg upload clinical-delivery -d knowingape`
image

- [x] dry-run with data delivery set as `fastq-analysis` and there's enough reads:`cg upload clinical-delivery -d knowingape`
```
----------------- UPLOAD -----------------
----------------- Clinical-delivery -----------------
Setting ticket to 123456
Setting customer_id to cust000
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/CaseName-RNA-5M_downsampled_fraser_top_hits_clinical.tsv already exists!
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/CaseName-RNA-5M_downsampled_fraser_top_hits_research.tsv already exists!
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/CaseName-RNA-5M_downsampled_outrider_top_hits_clinical.tsv already exists!
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/CaseName-RNA-5M_downsampled_outrider_top_hits_research.tsv already exists!
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/multiqc_report.html already exists!
File /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/multiqc_data.json already exists!
Linked 0 files for case knowingape
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX.cram to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.cram
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.cram exists, no hard link was made for file SampleNameACCXXXXX.cram
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX.cram.crai to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.cram.crai
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.cram.crai exists, no hard link was made for file SampleNameACCXXXXX.cram.crai
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX_junction.bed.gz to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX_junction.bed.gz
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX_junction.bed.gz exists, no hard link was made for file SampleNameACCXXXXX_junction.bed.gz
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX_junction.bed.gz.tbi to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX_junction.bed.gz.tbi
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX_junction.bed.gz.tbi exists, no hard link was made for file SampleNameACCXXXXX_junction.bed.gz.tbi
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX.transcripts.gtf to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.transcripts.gtf
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.transcripts.gtf exists, no hard link was made for file SampleNameACCXXXXX.transcripts.gtf
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX.annotated.gtf to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.annotated.gtf
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.annotated.gtf exists, no hard link was made for file SampleNameACCXXXXX.annotated.gtf
Hard link file /home/proj/stage/housekeeper-bundles/knowingape/2024-04-04/ACCXXXXX.ReadsPerGene.out.tab to /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.ReadsPerGene.out.tab
Warning: Path /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled/SampleNameACCXXXXX/SampleNameACCXXXXX.ReadsPerGene.out.tab exists, no hard link was made for file SampleNameACCXXXXX.ReadsPerGene.out.tab
There were 7 previously linked files and 0 were linked for sample ACCXXXXX, case knowingape
Setting ticket to 123456
Setting customer_id to cust000
Hard link file /home/proj/stage/housekeeper-bundles/ACCXXXXX/2024-04-02/XXXXX_L001_R1_001.fastq.gz to /PATH2CUSTOMER_INBOX/123456/SampleNameACCXXXXX/HT2H5DSX2_ACC9480A3_S311_L001_R1_001.fastq.gz
Warning: Path /PATH2CUSTOMER_INBOX/123456/SampleNameACCXXXXX/HT2H5DSX2_ACC9480A3_S311_L001_R1_001.fastq.gz exists, no hard link was made for file HT2H5DSX2_ACC9480A3_S311_L001_R1_001.fastq.gz
Hard link file /home/proj/stage/housekeeper-bundles/ACCXXXXX/2024-04-02/XXXX_L001_R2_001.fastq.gz to /PATH2CUSTOMER_INBOX/123456/SampleNameACCXXXXX/HT2H5DSX2_ACC9480A3_S311_L001_R2_001.fastq.gz
Warning: Path /PATH2CUSTOMER_INBOX/123456/SampleNameACCXXXXX/HT2H5DSX2_ACC9480A3_S311_L001_R2_001.fastq.gz exists, no hard link was made for file HT2H5DSX2_ACC9480A3_S311_L001_R2_001.fastq.gz
There were 2 previously linked files and 0 were linked for sample ACCXXXXX, case knowingape
Setting log dir to: /home/proj/stage/rsync/knowingape_240415_14_05_22_763532
Creating folder: /home/proj/stage/rsync/knowingape_240415_14_05_22_763532
Would have created path /home/proj/stage/rsync/knowingape_240415_14_05_22_763532, but this is a dry run
Write sbatch content to path /home/proj/stage/rsync/knowingape_240415_14_05_22_763532/knowingape_rsync.sh:
#! /bin/bash
#SBATCH --job-name=knowingape_rsync
#SBATCH --account=development
#SBATCH --ntasks=1
#SBATCH --mem=1G
#SBATCH --error=/home/proj/stage/rsync/knowingape_240415_14_05_22_763532/knowingape_rsync.stderr
#SBATCH --output=/home/proj/stage/rsync/knowingape_240415_14_05_22_763532/knowingape_rsync.stdout
#SBATCH --mail-type=FAIL
#SBATCH --mail-user=XXXX
#SBATCH --time=24:00:00
#SBATCH --qos=low
#SBATCH --exclude=XXXX

set -eu -o pipefail

log() {
    NOW=$(date +"%Y-%m-%dT%H:%M:%S")
    echo "[${NOW}] $*" 1>&2;
}

log "Running on: $(hostname)"


error() {

echo "Rsync failed"

    exit 1
}

trap error ERR


rsync -rvpL /PATH2CUSTOMER_INBOX/123456/SampleNameACCXXXXX DELIVERY_SERVER:/home/cust000/inbox/123456

rsync -rvpL /PATH2CUSTOMER_INBOX/123456/CaseName-RNA-5M_downsampled DELIVERY_SERVER:/home/cust000/inbox/123456



Submit sbatch /home/proj/stage/rsync/knowingape_240415_14_05_22_763532/knowingape_rsync.sh
Running command sbatch /home/proj/stage/rsync/knowingape_240415_14_05_22_763532/knowingape_rsync.sh
Dry run: process call will not be executed!!

Could not get slurm job number
Dry-run: it would write slurm jobs to /home/proj/stage/rsync/knowingape_240415_14_05_22_763532/slurm_job_ids.yaml
Transfer of case knowingape started with SLURM job id 123456
```

- [x] ` cg upload -c knowingape -r`. Make sure that upload is sucessful. Make sure that fastq files, case files and sample files were linked. No deliver report should be generated as it's not available yet.

image

- [x]  `cg upload auto --help`. Make sure tomte appears as a valid choice. 

## Review

- [ ] Tests executed by
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [ ] **MAJOR** - when you make incompatible API changes
- [ ] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Document in ...
- [ ] Deploy this branch on ...
- [ ] Inform to ...

### CODE REVIEW

File: cg/cli/upload/base.py [hunk 1]
```diff
@@ -77,12 +77,12 @@ def upload(context: click.Context, case_id: str | None, restart: bool):
...
         if Workflow.BALSAMIC in case.data_analysis:
             upload_api = BalsamicUploadAPI(config_object)
-        elif case.data_analysis == Workflow.RNAFUSION:
-            upload_api = RnafusionUploadAPI(config_object)
         elif case.data_analysis == Workflow.MIP_RNA:
             upload_api = MipRNAUploadAPI(config_object)
         elif case.data_analysis == Workflow.MICROSALT:
             upload_api = MicrosaltUploadAPI(config_object)
+        elif case.data_analysis in {Workflow.RNAFUSION, Workflow.TOMTE}:
+            upload_api = NfAnalysisUploadAPI(config_object)
```
Comment (diitaz93) [inline cg/cli/upload/base.py:85]: heads up for merge conflicts @sofstam in #3108. Maybe this class could be used for taxprofiler uploads too

**Review by diitaz93 — APPROVED** [review]
Great 💯

File: cg/cli/upload/base.py (hunk 1 above)
Comment (fevac) [reply cg/cli/upload/base.py:85]: woops! thanks for linking!

File: cg/constants/delivery.py [hunk 2]
```diff
@@ -169,6 +165,26 @@
...
+
+TOMTE_ANALYSIS_SAMPLE_TAGS: list[set[str]] = [
+    {AlignmentFileTag.CRAM},
+    {AlignmentFileTag.CRAM_INDEX},
+    {AnalysisTag.JUNCTION, AnalysisTag.BED},
+    {AnalysisTag.JUNCTION, AnalysisTag.BED_INDEX},
+    {AnalysisTag.STRINGTIE, AnalysisTag.ASSEMBLY},
+    {AnalysisTag.GFFCOMPARE},
+    {AnalysisTag.GENE_COUNTS},
+]
```
Comment (ivadym) [inline cg/constants/delivery.py:185]:
```suggestion
TOMTE_ANALYSIS_CASE_TAGS: list[set[str]] = [{HermesFileTag.CLINICAL_DELIVERY}]

TOMTE_ANALYSIS_SAMPLE_TAGS: list[set[str]] = [{HermesFileTag.CLINICAL_DELIVERY}]
```

File: cg/cli/upload/base.py (hunk 1 above)
Comment (sofstam) [reply cg/cli/upload/base.py:85]: Sounds good, thanks for the heads-up @diitaz93

**Review by ivadym — COMMENTED** [review]
Well done, but let's use the `clinical-delivery` tag instead 👍 

And I think we forgot to add tomte to the list of protected tags in the store PR (`long-term-storage` tag)

File: cg/constants/delivery.py (hunk 2 above)
Comment (fevac) [reply cg/constants/delivery.py]: added

**Comment by fevac** [comment]
- deployed

image

image

**Pull request merged**